<a href="https://colab.research.google.com/github/NguyenSang2003/Machine-Learning/blob/main/Lab_6_21130512_NguyenVanSang.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# This lab is to deal with classification task using **Random Forests** and **Naïve Bayes** algorithms with/without **Feature Selection**.

*   **Deadline: 23:59, 15/04/2024**



# Import libraries

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn import preprocessing
from sklearn import datasets
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn import metrics
from sklearn.impute import SimpleImputer
from sklearn.linear_model import LinearRegression,LogisticRegression
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.naive_bayes import GaussianNB
from sklearn import svm
from sklearn.svm import SVC
from sklearn.preprocessing import OneHotEncoder
from sklearn.tree import DecisionTreeClassifier
from sklearn.feature_selection import SelectKBest, f_classif
from sklearn.preprocessing import LabelEncoder

In [ ]:
from google.colab import drive
drive.mount('/content/gdrive')
%cd '/content/gdrive/MyDrive/Data_Labs_ML/Lab6'

Mounted at /content/gdrive
/content/gdrive/MyDrive/Data_Labs_ML/Lab6


#Task 1.
Task 1. Compare the performance of selected classification algorithms including **Random forest**, **NaiveBayes**, and **SVM** with **FASHION** dataset based on **accuracy, precision, recall, f1** measures according to **without using selection feature** and **using selection feature**.



In [ ]:
# Đọc dữ liệu từ tập train và test
train_data = pd.read_csv("fashion_train.csv")
test_data = pd.read_csv("fashion_test.csv")

# Tách đặc trưng và nhãn từ tập dữ liệu train
X_train = train_data.drop('y', axis=1)
y_train = train_data['y']

# Tách đặc trưng và nhãn từ tập dữ liệu test
X_test = test_data.drop('y', axis=1)
y_test = test_data['y']

train_data.tail()

V1  V2  V3  V4  V5  V6  V7  V8  V9  V10  ...  V776  V777  V778  V779  \
995   0   0   0   0   0   0   0   0   0    0  ...     0     0     0     0   
996   0   0   0   0   0   0   0   0   0    0  ...     0     0     0     0   
997   0   0   0   0   0   0   0   0   0    0  ...   197   195     3     0   
998   0   0   0   0   0   0   0   0   0    0  ...     0     0   180   247   
999   0   0   1   0   1   0   0   0   0   77  ...     0     0     0     0   

     V780  V781  V782  V783  V784  y  
995     0     0     0     0     0  5  
996     0     0     0     0     0  9  
997     3     0     0     0     0  3  
998   210     0     0     0     0  4  
999   172   180   218    70     0  2  

[5 rows x 785 columns]

In [ ]:
# task 1 mà không sử dụng selection
print("So sánh hiệu suất của các thuật toán phân loại đã chọn (Random Forest, Naive Bayes và SVM)")
print("trên tập dữ liệu FASHION, mà không sử dụng lựa chọn đặc trưng: ")

# Random Forest
rf_clf = RandomForestClassifier(n_estimators=100)
rf_clf.fit(X_train, y_train)
rf_y_pred = rf_clf.predict(X_test)

print("\nRandom Forest:")
print("Accuracy:", metrics.accuracy_score(y_test, rf_y_pred))
print("Precision:", metrics.precision_score(y_test, rf_y_pred, average='macro'))
print("Recall:", metrics.recall_score(y_test, rf_y_pred, average='macro'))
print("F1 measure:", metrics.f1_score(y_test, rf_y_pred, average='macro'))

# Naive Bayes
nb_model = GaussianNB()
nb_model.fit(X_train, y_train)
nb_y_pred = nb_model.predict(X_test)

print("\nNaive Bayes:")
print("Accuracy:", metrics.accuracy_score(y_test, nb_y_pred))
print("Precision:", metrics.precision_score(y_test, nb_y_pred, average='macro'))
print("Recall:", metrics.recall_score(y_test, nb_y_pred, average='macro'))
print("F1 measure:", metrics.f1_score(y_test, nb_y_pred, average='macro'))

# SVM
svm_linear = SVC(kernel='linear')
svm_linear.fit(X_train, y_train)
svm_linear_pred = svm_linear.predict(X_test)

print("\nSVM:")
print("Accuracy:", metrics.accuracy_score(y_test, svm_linear_pred))
print("Precision:", metrics.precision_score(y_test, svm_linear_pred, average='macro'))
print("Recall:", metrics.recall_score(y_test, svm_linear_pred, average='macro'))
print("F1 measure:", metrics.f1_score(y_test, svm_linear_pred, average='macro'))


So sánh hiệu suất của các thuật toán phân loại đã chọn (Random Forest, Naive Bayes và SVM)
trên tập dữ liệu FASHION, mà không sử dụng lựa chọn đặc trưng: 

Random Forest:
Accuracy: 0.81
Precision: 0.8035298606614909
Recall: 0.8070363257780617
F1 measure: 0.8014778138282284

Naive Bayes:
Accuracy: 0.556
Precision: 0.5788628371304589
Recall: 0.559496772854223
F1 measure: 0.5256907025966638

SVM:
Accuracy: 0.809
Precision: 0.8117729613602249
Recall: 0.8080638899207333
F1 measure: 0.8054464844513045


In [ ]:
# Task 1 sử dụng selection
print("So sánh hiệu suất của các thuật toán phân loại đã chọn (Random Forest, Naive Bayes và SVM)")
print("trên tập dữ liệu FASHION, sử dụng lựa chọn đặc trưng: ")

# Áp dụng PCA để giảm chiều dữ liệu
pca = PCA(n_components=3)
X_new_train = pca.fit_transform(X_train)
X_new_test = pca.fit_transform(X_test)

# Random Forest
rf_clf = RandomForestClassifier(n_estimators=100)
rf_clf.fit(X_new_train, y_train)
rf_y_pred = rf_clf.predict(X_new_test)

print("\nRandom Forest:")
print("Accuracy:", metrics.accuracy_score(y_test, rf_y_pred))
print("Precision:", metrics.precision_score(y_test, rf_y_pred, average='macro'))
print("Recall:", metrics.recall_score(y_test, rf_y_pred, average='macro'))
print("F1 measure:", metrics.f1_score(y_test, rf_y_pred, average='macro'))

# Naive Bayes
nb_model = GaussianNB()
nb_model.fit(X_new_train, y_train)
nb_y_pred = nb_model.predict(X_new_test)

print("\nNaive Bayes:")
print("Accuracy:", metrics.accuracy_score(y_test, nb_y_pred))
print("Precision:", metrics.precision_score(y_test, nb_y_pred, average='macro'))
print("Recall:", metrics.recall_score(y_test, nb_y_pred, average='macro'))
print("F1 measure:", metrics.f1_score(y_test, nb_y_pred, average='macro'))

# SVM
svm_linear = SVC(kernel='linear')
svm_linear.fit(X_new_train, y_train)
svm_linear_pred = svm_linear.predict(X_new_test)

print("\nSVM:")
print("Accuracy:", metrics.accuracy_score(y_test, svm_linear_pred))
print("Precision:", metrics.precision_score(y_test, svm_linear_pred, average='macro'))
print("Recall:", metrics.recall_score(y_test, svm_linear_pred, average='macro'))
print("F1 measure:", metrics.f1_score(y_test, svm_linear_pred, average='macro'))


So sánh hiệu suất của các thuật toán phân loại đã chọn (Random Forest, Naive Bayes và SVM)
trên tập dữ liệu FASHION, sử dụng lựa chọn đặc trưng: 

Random Forest:
Accuracy: 0.576
Precision: 0.5652954794473782
Recall: 0.5683456909261677
F1 measure: 0.56184603485676

Naive Bayes:
Accuracy: 0.581
Precision: 0.5816308025724025
Recall: 0.5792108522099866
F1 measure: 0.5673237100180284

SVM:
Accuracy: 0.58
Precision: 0.5583143092326797
Recall: 0.5755993157494911
F1 measure: 0.5291728458766034


#Task 2.
For given bank dataset (**bank.csv**) having the following attributes :
1.	**age** (numeric)
2.	**job** : type of job (categorical: 'admin.','blue-collar','entrepreneur','housemaid','management','retired','self-employed','services','student','technician','unemployed','unknown')
3.	**marital** : marital status (categorical: 'divorced','married','single','unknown'; note: 'divorced' means divorced or widowed)
4.	**education** (categorical: 'basic.4y','basic.6y','basic.9y','high.school','illiterate','professional.course','university.degree','unknown')
5.	**default**: has credit in default? (categorical: 'no','yes','unknown')
6.	**housing**: has housing loan? (categorical: 'no','yes','unknown')
7.	**loan**: has personal loan? (categorical: 'no','yes','unknown')
8.	**contact**: contact communication type (categorical: 'cellular','telephone')
9.	**month**: last contact month of year (categorical: 'jan', 'feb', 'mar', ..., 'nov', 'dec')
10.	**day_of_week**: last contact day of the week (categorical: 'mon','tue','wed','thu','fri')
11.	**duration**: last contact duration, in seconds (numeric). Important note: this attribute highly affects the output target (e.g., if duration=0 then y='no'). Yet, the duration is not known before a call is performed. Also, after the end of the call y is obviously known. Thus, this input should only be included for benchmark purposes and should be discarded if the intention is to have a realistic predictive model.
12.	**campaign**: number of contacts performed during this campaign and for this client (numeric, includes last contact)
13.	**pdays**: number of days that passed by after the client was last contacted from a previous campaign (numeric; 999 means client was not previously contacted)
14.	**previous**: number of contacts performed before this campaign and for this client (numeric)
15.	**poutcome**: outcome of the previous marketing campaign (categorical: 'failure','nonexistent','success')
Output variable (desired target):
16.	**y**. has the client subscribed a term deposit? (binary: 'yes','no')



In [55]:
# Đọc dữ liệu từ tập train và test
data_bank= pd.read_csv("bank.csv")

data_bank.tail()

age          job  marital  education default  balance housing loan  \
11157   33  blue-collar   single    primary      no        1     yes   no   
11158   39     services  married  secondary      no      733      no   no   
11159   32   technician   single  secondary      no       29      no   no   
11160   43   technician  married  secondary      no        0      no  yes   
11161   34   technician  married  secondary      no        0      no   no   

        contact  day month  duration  campaign  pdays  previous poutcome  \
11157  cellular   20   apr       257         1     -1         0  unknown   
11158   unknown   16   jun        83         4     -1         0  unknown   
11159  cellular   19   aug       156         2     -1         0  unknown   
11160  cellular    8   may         9         2    172         5  failure   
11161  cellular    9   jul       628         1     -1         0  unknown   

      deposit  
11157      no  
11158      no  
11159      no  
11160      no  
11161      no

#2.1. Apply StandardScaler() function to columns that contains numerical data ('age', 'balance', 'day', 'campaign', 'pdays', 'previous')

In [ ]:
# Các cột chứa dữ liệu số cần được chuẩn hóa
numeric_columns = ['age', 'balance', 'day', 'campaign', 'pdays', 'previous']

# Khởi tạo StandardScaler
scaler = StandardScaler()

# Chuẩn hóa các cột dữ liệu số
data_bank[numeric_columns] = scaler.fit_transform(data_bank[numeric_columns])

# In ra 5 dòng cuối của dữ liệu để kiểm tra
print(data_bank.tail())

            age          job  marital  education default   balance housing  \
11157 -0.691015  blue-collar   single    primary      no -0.473616     yes   
11158 -0.187357     services  married  secondary      no -0.246658      no   
11159 -0.774958   technician   single  secondary      no -0.464934      no   
11160  0.148416   technician  married  secondary      no -0.473926      no   
11161 -0.607072   technician  married  secondary      no -0.473926      no   

      loan   contact       day month  duration  campaign     pdays  previous  \
11157   no  cellular  0.515650   apr       257 -0.554168 -0.481184 -0.363260   
11158   no   unknown  0.040612   jun        83  0.547981 -0.481184 -0.363260   
11159   no  cellular  0.396891   aug       156 -0.186785 -0.481184 -0.363260   
11160  yes  cellular -0.909466   may         9 -0.186785  1.109571  1.818332   
11161   no  cellular -0.790707   jul       628 -0.554168 -0.481184 -0.363260   

      poutcome deposit  
11157  unknown      no  


#2.2. Apply Encode Categorical Value (OneHotEncoder) to transfrom categorical data to numerical data ('job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'month', 'poutcome')

In [ ]:
data_bank.columns

Index(['age', 'job', 'marital', 'education', 'default', 'balance', 'housing',
       'loan', 'contact', 'day', 'month', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'deposit'],
      dtype='object')

In [ ]:
# Các cột chứa dữ liệu hạng mục cần được mã hóa
categorical_columns = ['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'month', 'poutcome']

# Khởi tạo OneHotEncoder
encoder = OneHotEncoder(drop='first', sparse=False)

# Mã hóa các cột dữ liệu hạng mục
encoded_data = encoder.fit_transform(data_bank[categorical_columns])

# Tạo DataFrame mới từ dữ liệu đã mã hóa
encoded_df = pd.DataFrame(encoded_data, columns=encoder.get_feature_names_out(categorical_columns))

# Loại bỏ các cột gốc đã được mã hóa
data = data_bank.drop(columns=categorical_columns)

# Ghép DataFrame mới đã mã hóa vào DataFrame gốc
data = pd.concat([data, encoded_df], axis=1)

# In ra 5 dòng cuối cùng của dữ liệu để kiểm tra
print(data.tail())

            age   balance       day  duration  campaign     pdays  previous  \
11157 -0.691015 -0.473616  0.515650       257 -0.554168 -0.481184 -0.363260   
11158 -0.187357 -0.246658  0.040612        83  0.547981 -0.481184 -0.363260   
11159 -0.774958 -0.464934  0.396891       156 -0.186785 -0.481184 -0.363260   
11160  0.148416 -0.473926 -0.909466         9 -0.186785  1.109571  1.818332   
11161 -0.607072 -0.473926 -0.790707       628 -0.554168 -0.481184 -0.363260   

      deposit  job_blue-collar  job_entrepreneur  ...  month_jul  month_jun  \
11157      no              1.0               0.0  ...        0.0        0.0   
11158      no              0.0               0.0  ...        0.0        1.0   
11159      no              0.0               0.0  ...        0.0        0.0   
11160      no              0.0               0.0  ...        0.0        0.0   
11161      no              0.0               0.0  ...        1.0        0.0   

       month_mar  month_may  month_nov  month_oct 

/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_encoders.py:868: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


#2.3. Apply **Decision tree, Logistic Regression, Random forest, kNN, NaïveBayes, SVM** algorithms to the preproceed dataset in the previous steps. Then compare the obtained results using **accuracy, precision, recall, f1** measures.

In [ ]:
# Chia tập dữ liệu thành tập huấn luyện và tập kiểm tra
X = data.drop(columns=['deposit'])  # Features
y = data['deposit']  # Target variable
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Khởi tạo các mô hình
models = {
    'Decision Tree': DecisionTreeClassifier(),
    'Logistic Regression': LogisticRegression(),
    'Random Forest': RandomForestClassifier(),
    'kNN': KNeighborsClassifier(),
    'Naïve Bayes': GaussianNB(),
    'SVM': SVC()
}

# Huấn luyện và đánh giá các mô hình
models_results = {}
for name, model in models.items():
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred, pos_label='yes')
    recall = recall_score(y_test, y_pred, pos_label='yes')
    f1 = f1_score(y_test, y_pred, pos_label='yes')
    models_results[name] = [accuracy, precision, recall, f1]

# Hiển thị kết quả
results_df = pd.DataFrame(models_results, index=['Accuracy', 'Precision', 'Recall', 'F1 Score'])
print(results_df)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


           Decision Tree  Logistic Regression  Random Forest       kNN  \
Accuracy        0.791760             0.810569       0.835199  0.749664   
Precision       0.785038             0.809615       0.808473  0.740530   
Recall          0.776945             0.789128       0.858482  0.732896   
F1 Score        0.780970             0.799241       0.832727  0.736693   

           Naïve Bayes       SVM  
Accuracy      0.723242  0.706673  
Precision     0.824891  0.707661  
Recall        0.534208  0.657919  
F1 Score      0.648464  0.681884  


#2.4. Using a selection feature technique to above dataset, then compare the classification results with those in Task 2.3.

In [ ]:
# Chia dữ liệu thành tập huấn luyện và tập kiểm tra
X_train, X_test, y_train, y_test = train_test_split(data.drop('deposit', axis=1), data['deposit'], test_size=0.2, random_state=42)

# Sử dụng PCA để giảm chiều dữ liệu
pca = PCA(n_components=10)  # Chọn số lượng thành phần chính phù hợp
X_train_pca = pca.fit_transform(X_train)
X_test_pca = pca.transform(X_test)

# Áp dụng các mô hình phân loại đã chọn từ Task 2.3
models = {
    "Decision Tree": DecisionTreeClassifier(),
    "Logistic Regression": LogisticRegression(),
    "Random Forest": RandomForestClassifier(),
    "kNN": KNeighborsClassifier(),
    "Naive Bayes": GaussianNB(),
    "SVM": SVC()
}

# Lặp qua từng mô hình và đánh giá kết quả phân loại
results_pca = {}
for name, model in models.items():
    model.fit(X_train_pca, y_train)
    y_pred = model.predict(X_test_pca)
    accuracy = accuracy_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred, average='weighted')
    recall = recall_score(y_test, y_pred, average='weighted')
    f1 = f1_score(y_test, y_pred, average='weighted')
    results_pca[name] = [accuracy, precision, recall, f1]

# In ra kết quả
print("Results after PCA:")
print("===================================")
print("{:<20} {:<10} {:<10} {:<10} {:<10}".format("Algorithm", "Accuracy", "Precision", "Recall", "F1 Score"))
for name, scores in results_pca.items():
    print("{:<20} {:<10.4f} {:<10.4f} {:<10.4f} {:<10.4f}".format(name, scores[0], scores[1], scores[2], scores[3]))

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Results after PCA:
Algorithm            Accuracy   Precision  Recall     F1 Score  
Decision Tree        0.7300     0.7298     0.7300     0.7298    
Logistic Regression  0.7824     0.7824     0.7824     0.7821    
Random Forest        0.8025     0.8037     0.8025     0.8026    
kNN                  0.7519     0.7518     0.7519     0.7518    
Naive Bayes          0.7344     0.7440     0.7344     0.7298    
SVM                  0.7080     0.7083     0.7080     0.7071    


#Task 3.
For a given dataset (**mobile price classification**), perform feature selection techniques and then compare the performance of selected classification algorithms (**Decision Tree, kNN, Logistic Regression, SVM, Random Forest, and NaiveBayes**) based on **accuracy, precision, recall, and f1** measures.


In [ ]:
# Đọc dữ liệu từ tập train và test
data = pd.read_csv("mobile.csv")

data.tail()

battery_power  blue  clock_speed  dual_sim  fc  four_g  int_memory  \
1995            794     1          0.5         1   0       1           2   
1996           1965     1          2.6         1   0       0          39   
1997           1911     0          0.9         1   1       1          36   
1998           1512     0          0.9         0   4       1          46   
1999            510     1          2.0         1   5       1          45   

      m_dep  mobile_wt  n_cores  ...  px_height  px_width   ram  sc_h  sc_w  \
1995    0.8        106        6  ...       1222      1890   668    13     4   
1996    0.2        187        4  ...        915      1965  2032    11    10   
1997    0.7        108        8  ...        868      1632  3057     9     1   
1998    0.1        145        5  ...        336       670   869    18    10   
1999    0.9        168        6  ...        483       754  3919    19     4   

      talk_time  three_g  touch_screen  wifi  price_range  
1995         19        1             1     0            0  
1996         16        1             1     1            2  
1997          5        1             1     0            3  
1998         19        1             1     1            0  
1999          2        1             1     1            3  

[5 rows x 21 columns]

In [ ]:
X = data.drop('price_range', axis=1)
y = data['price_range']

# Chia dữ liệu thành tập huấn luyện và tập kiểm tra
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Chuẩn hóa dữ liệu
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Lựa chọn đặc trưng
selector = SelectKBest(f_classif, k=10)
X_train_selected = selector.fit_transform(X_train_scaled, y_train)
X_test_selected = selector.transform(X_test_scaled)

# Khởi tạo các mô hình phân loại
classifiers = {
    'Decision Tree': DecisionTreeClassifier(),
    'kNN': KNeighborsClassifier(),
    'Logistic Regression': LogisticRegression(),
    'SVM': SVC(),
    'Random Forest': RandomForestClassifier(),
    'Naive Bayes': GaussianNB()
}

# Huấn luyện và đánh giá các mô hình phân loại
results = {}
for clf_name, clf in classifiers.items():
    clf.fit(X_train_selected, y_train)
    y_pred = clf.predict(X_test_selected)
    accuracy = accuracy_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred, average='weighted')
    recall = recall_score(y_test, y_pred, average='weighted')
    f1 = f1_score(y_test, y_pred, average='weighted')
    results[clf_name] = {'Accuracy': accuracy, 'Precision': precision, 'Recall': recall, 'F1 Score': f1}

# In kết quả
results_df = pd.DataFrame.from_dict(results, orient='index')
print(results_df)

                     Accuracy  Precision  Recall  F1 Score
Decision Tree          0.8400   0.843337  0.8400  0.840548
kNN                    0.6675   0.693893  0.6675  0.675326
Logistic Regression    0.9850   0.985337  0.9850  0.984998
SVM                    0.9200   0.923197  0.9200  0.920742
Random Forest          0.9200   0.920874  0.9200  0.920218
Naive Bayes            0.8150   0.825944  0.8150  0.818061


#Task 4.
For a given dataset (named **mushroom.csv**) with the following information:

*   Attribute Information: (**classes**: edible=e, poisonous=p)
*   **cap-shape**: bell=b,conical=c,convex=x,flat=f, knobbed=k,sunken=s
*   **cap-surface**: fibrous=f,grooves=g,scaly=y,smooth=s
*   **cap-color**: brown=n,buff=b,cinnamon=c,gray=g,green=r,pink=p,purple=u,red=e,white=w,yellow=y
*   **bruises**: bruises=t,no=f
*   **odor**: almond=a,anise=l,creosote=c,fishy=y,foul=f,musty=m,none=n,pungent=p,spicy=s
*   **gill-attachment**: attached=a,descending=d,free=f,notched=n
*   **gill-spacing**: close=c,crowded=w,distant=d
*   **gill-size**: broad=b,narrow=n
*   **gill-color**: black=k,brown=n,buff=b,chocolate=h,gray=g, green=r,orange=o,pink=p,purple=u,red=e,white=w,yellow=y
*   **stalk-shape**: enlarging=e,tapering=t
*   **stalk-root**: bulbous=b,club=c,cup=u,equal=e,rhizomorphs=z,rooted=r,missing=?
*   **stalk-surface-above-ring**: fibrous=f,scaly=y,silky=k,smooth=s
*   **stalk-surface-below-ring**: fibrous=f,scaly=y,silky=k,smooth=s
*   **stalk-color-above-ring**: brown=n,buff=b,cinnamon=c,gray=g,orange=o,pink=p,red=e,white=w,yellow=y
*   **stalk-color-below-ring**: brown=n,buff=b,cinnamon=c,gray=g,orange=o,pink=p,red=e,white=w,yellow=y
*   **veil-type**: partial=p,universal=u
*   **veil-color**: brown=n,orange=o,white=w,yellow=y
*   **ring-number**: none=n,one=o,two=t
*   **ring-type**: cobwebby=c,evanescent=e,flaring=f,large=l,none=n,pendant=p,sheathing=s,zone=z
*   **spore-print-color**: black=k,brown=n,buff=b,chocolate=h,green=r,orange=o,purple=u,white=w,yellow=y
*   **population**: abundant=a,clustered=c,numerous=n,scattered=s,several=v,solitary=y
*   **habitat**: grasses=g,leaves=l,meadows=m,paths=p,urban=u,waste=w,woods=d

In [ ]:
# Đọc dữ liệu từ file
data = pd.read_csv("mushrooms.csv")

data.tail()

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
8119     e         k           s         n       f    n               a   
8120     e         x           s         n       f    n               a   
8121     e         f           s         n       f    n               a   
8122     p         k           y         n       f    y               f   
8123     e         x           s         n       f    n               a   

     gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
8119            c         b          y  ...                        s   
8120            c         b          y  ...                        s   
8121            c         b          n  ...                        s   
8122            c         n          b  ...                        k   
8123            c         b          y  ...                        s   

     stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
8119                      o                      o         p          o   
8120                      o                      o         p          n   
8121                      o                      o         p          o   
8122                      w                      w         p          w   
8123                      o                      o         p          o   

     ring-number ring-type spore-print-color population habitat  
8119           o         p                 b          c       l  
8120           o         p                 b          v       l  
8121           o         p                 b          c       l  
8122           o         e                 w          v       l  
8123           o         p                 o          c       l  

[5 rows x 23 columns]

#Task 4.1. Apply approriate reprocessing techniques for the mushroom dataset

In [ ]:
# Mã hóa biến phân loại
label_encoder = LabelEncoder()
encoded_data = data.apply(label_encoder.fit_transform)

# In ra 5 dòng đầu tiên của dữ liệu đã mã hóa
print(encoded_data.head())

   class  cap-shape  cap-surface  cap-color  bruises  odor  gill-attachment  \
0      1          5            2          4        1     6                1   
1      0          5            2          9        1     0                1   
2      0          0            2          8        1     3                1   
3      1          5            3          8        1     6                1   
4      0          5            2          3        0     5                1   

   gill-spacing  gill-size  gill-color  ...  stalk-surface-below-ring  \
0             0          1           4  ...                         2   
1             0          0           4  ...                         2   
2             0          0           5  ...                         2   
3             0          1           5  ...                         2   
4             1          0           4  ...                         2   

   stalk-color-above-ring  stalk-color-below-ring  veil-type  veil-color  \
0         

#Task 4.2. Apply classification algorithms (**Decision Tree, Logistic Regression, SVM, kNN, Random Forest, NaiveBayes**) to the reprocessed dataset

In [ ]:
# Chia dữ liệu thành features (X) và target (y)
X = encoded_data.drop('class', axis=1)
y = encoded_data['class']

# Chia dữ liệu thành tập huấn luyện và tập kiểm tra
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Khởi tạo và huấn luyện các mô hình phân loại
models = {
    'Decision Tree': DecisionTreeClassifier(),
    'Logistic Regression': LogisticRegression(),
    'SVM': SVC(),
    'kNN': KNeighborsClassifier(),
    'Random Forest': RandomForestClassifier(),
    'Naive Bayes': GaussianNB()
}

results = {}
for name, model in models.items():
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred)
    recall = recall_score(y_test, y_pred)
    f1 = f1_score(y_test, y_pred)
    results[name] = {'Accuracy': accuracy, 'Precision': precision, 'Recall': recall, 'F1 Score': f1}

# In kết quả
results_df = pd.DataFrame.from_dict(results, orient='index')
print(results_df)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


                     Accuracy  Precision    Recall  F1 Score
Decision Tree        1.000000   1.000000  1.000000  1.000000
Logistic Regression  0.947077   0.942748  0.947570  0.945153
SVM                  0.992615   0.998705  0.985934  0.992278
kNN                  0.996308   0.992386  1.000000  0.996178
Random Forest        1.000000   1.000000  1.000000  1.000000
Naive Bayes          0.921846   0.909887  0.929668  0.919671


#Task 4.3. Apply an feature selection method to dataset obtained in Task 4.1, then apply classification algorithm to the obtained dataset. Compare the performance of classification models before and after applying the feature selection technique in term of **accuracy, precision, recall, and F1** measures

In [ ]:
# Áp dụng phương pháp lựa chọn đặc trưng
selector = SelectKBest(score_func=f_classif, k=10)
X_train_selected = selector.fit_transform(X_train, y_train)
X_test_selected = selector.transform(X_test)

# Khởi tạo và huấn luyện các mô hình phân loại trên tập dữ liệu đã lựa chọn đặc trưng
models_results = {}
for name, model in models.items():
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred)
    recall = recall_score(y_test, y_pred)
    f1 = f1_score(y_test, y_pred)
    models_results[name] = {'Accuracy Before': accuracy, 'Precision Before': precision, 'Recall Before': recall, 'F1 Score Before': f1}

# Khởi tạo và huấn luyện các mô hình phân loại trên tập dữ liệu đã lựa chọn đặc trưng
for name, model in models.items():
    model.fit(X_train_selected, y_train)
    y_pred = model.predict(X_test_selected)
    accuracy = accuracy_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred)
    recall = recall_score(y_test, y_pred)
    f1 = f1_score(y_test, y_pred)
    models_results[name]['Accuracy After'] = accuracy
    models_results[name]['Precision After'] = precision
    models_results[name]['Recall After'] = recall
    models_results[name]['F1 Score After'] = f1

# Chuyển kết quả thành DataFrame
results_df = pd.DataFrame.from_dict(models_results, orient='index')

print(results_df)

/usr/local/lib/python3.10/dist-packages/sklearn/feature_selection/_univariate_selection.py:112: UserWarning: Features [15] are constant.
  warnings.warn("Features %s are constant." % constant_features_idx, UserWarning)
/usr/local/lib/python3.10/dist-packages/sklearn/feature_selection/_univariate_selection.py:113: RuntimeWarning: invalid value encountered in divide
  f = msb / msw
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


                     Accuracy Before  Precision Before  Recall Before  \
Decision Tree               1.000000          1.000000       1.000000   
Logistic Regression         0.947077          0.942748       0.947570   
SVM                         0.992615          0.998705       0.985934   
kNN                         0.996308          0.992386       1.000000   
Random Forest               1.000000          1.000000       1.000000   
Naive Bayes                 0.921846          0.909887       0.929668   

                     F1 Score Before  Accuracy After  Precision After  \
Decision Tree               1.000000        0.995692         0.996159   
Logistic Regression         0.945153        0.939692         0.946475   
SVM                         0.992278        0.969846         0.979085   
kNN                         0.996178        0.995692         0.996159   
Random Forest               1.000000        0.995692         0.996159   
Naive Bayes                 0.919671        0.9009

#Finally,
Save a copy in your Github. Remember renaming the notebook.